Mathlib/LinearAlgebra/Matrix/FixedDetMatrices.lean

```diff
@@ -52,4 +53,205 @@ lemma smul_coe (m : R) (g : SpecialLinearGroup n R) (A : FixedDetMatrix n R m) :
     (g • A).1 = g * A.1 := by
   rw [smul_def]
 
+section IntegralFixedDetMatrices
+
+local notation:1024 "Δ" m : 1024 => (FixedDetMatrix (Fin 2) ℤ m)
+
+variable {m : ℤ}
+
+/--Set of representatives for the orbits under `S` and `T`. -/
+def reps (m : ℤ) : Set (Δ m) :=
+  { A : Δ m | (A.1 1 0) = 0 ∧ 0 < A.1 0 0 ∧ 0 ≤ A.1 0 1 ∧ |(A.1 0 1)| < |(A.1 1 1)|}
+
+/--Reduction step for matrices in `Δ m` which moves the matrices towards `reps`.-/
+def reduce_step (A : Δ m) : Δ m := S • (T ^ (-(A.1 0 0 / A.1 1 0))) • A
+
+private lemma reduce_aux {A : Δ m} (h : |(A.1 1 0)| ≠ 0) :
+    |((reduce_step A).1 1 0)| < |(A.1 1 0)| := by
+  simp_rw [reduce_step, smul_coe, coe_T_zpow, S]
+  simpa only [Int.reduceNeg, cons_mul, Nat.succ_eq_add_one, Nat.reduceAdd, empty_mul,
+    Equiv.symm_apply_apply, vecMul_cons, vecHead, cons_val_zero, zero_smul, vecTail,
+    Function.comp_apply, Fin.succ_zero_eq_one, cons_val_one, cons_val_fin_one, neg_smul, one_smul,
+    empty_vecMul, add_zero, zero_add, of_apply, cons_val', Pi.neg_apply, vecMul, cons_dotProduct,
+    zero_mul, one_mul, dotProduct_empty, mul_comm, mul_neg, ← Int.sub_eq_add_neg,
+    ← Int.emod_def (A.1 0 0) (A.1 1 0), empty_val',
+    abs_eq_self.mpr (Int.emod_nonneg (A.1 0 0) (abs_ne_zero.mp h))]
+  using Int.emod_lt (A.1 0 0) (abs_ne_zero.mp h)
+
+/--Reduction lemma for integral FixedDetMatrices. -/
+@[elab_as_elim]
+def reduce_rec {C : Δ m → Sort*} (h0 : ∀ A : Δ m, |(A.1 1 0)| = 0 → C A)
```

Comment:
Why do you want an absolute value in the statement of `h0` and `h1`?
I understand this might be useful for the proof, but is it also the best for applications?

Reply:
yeah thats a good point, I think its just a leftover from some old stuff. I've updated it now! thanks!